New feature - Solar System ephemerides from the Minor Planet Center

mkelley opened the pull request:
This is an update to the current Minor Planet Center module to add the ability to retrieve comet and asteroid ephemerides from the MPC, as described in Issue #1075 .  The MPC's ephemeris service API is undocumented, but the web-service documentation states that script access is allowed, and that the HTML document should be used as a template.

The web-service interface is located at https://minorplanetcenter.net/iau/MPEph/MPEph.html

The documentation is at https://minorplanetcenter.net/iau/info/MPES.pdf

This first commit is a proposed docstring.

mkelley commented:
I've made a few choices that warrant some discussion.  The main question, do we try to reproduce the MPC service as closely as possible, or do we make small changes so that it more closely aligns with the astropy ecosystem?  I've leaned more into the second part:
1. Ephemerides are limited to three choices:
   1. full sexagesimal (RA precision to 0.1", Dec to 1"),
   1. heliocentric position and velocity vector, and
   1. geocentric position vector.

   There are three additional options available:

   1. truncated sexagesimal (RA to 0.01', Dec to 0.1'),
   1. RA and Dec in decimal format, and
   1. heliocentric position vector (without velocity).

   My assumption was that we only need one of each type, and with `Angle` and `Table` column formats, the coordinates can be displayed in whatever format the user wants after the table is returned. So, we could restore all six of the MPCs formats, we could add a keyword argument to set the column format for the user, or just return the table as is.
2. Proper motion can be returned as arcsec/s, arcsec/min, arcsec/hr, or deg/day.  With units, it is straightforward to convert between these and others.  We could add a keyword argument to do this for the user.
3. Azimuths may be returned measuring east of north, or west of south.  Astropy has an `AltAz` coordinate frame that uses east of north convention.  To avoid confusion, I've limited the MPC output to east of north.

It is also possible to return orbital elements.  I've assumed this will be a separate method.

keflavich commented:
I like the "keyword arguments to specify formats" approach.  Can you confirm, though, that astropy is capable of converting between all the various formats?  As long as that's true, and we include examples in the docs, I think it's better than the 'return all possible formats' approach.

mommermi commented:
@mkelley, this looks good to me - but probably @dmcollom should have a look at this, too.

I also fully agree with your points 1-3 and @keflavich.

dmcollom commented:
@mkelley, this looks like a good addition--didn't even know this service existed, so thanks for starting to add it.

I'm a software engineer, not an astronomer, so I'll defer to the astropy maintainers on the interface and business logic decisions. That said, just a couple early things from a software perspective:

1- Your imports and constant definitions should be moved to the top of the file.
2- There should be some sort of validation message in the case that proper motion/ephemeris type are invalid, instead of the current unchecked exception.

Looking forward to seeing the finished product.

migueldvb commented:
This looks very good. I also like the option to convert the ephemeris output units and format between all the various ephemeris output using astropy. Just a few minor comments.

-   In the `step` optional argument indicate that the default value is 6 hours (or                  
something different if the `query_ephemeris_async` method will set a different default).
- the module could include the file listing the observatory data to allow the user provide a matching observatory name as the location, or to query the observatory list to find the code: https://minorplanetcenter.net/iau/lists/ObsCodes.html
- Use 4 spaces before the description of the function arguments:
```
    epoch : str
        The date/time of reference for the current orbital parameters.                    
  ```

pep8speaks commented:
Hello @mkelley! Thanks for updating the PR.

 - In the file [`astroquery/mpc/core.py`](https://github.com/astropy/astroquery/blob/6d5b5ebe5dfb680dc9219e860c57b3d26d7e6165/astroquery/mpc/core.py), following are the PEP8 issues :

> [Line 542:17](https://github.com/astropy/astroquery/blob/6d5b5ebe5dfb680dc9219e860c57b3d26d7e6165/astroquery/mpc/core.py#L542): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator

mkelley commented:
Thanks all.  I've revised the draft, accounting for all your comments.  I've also added a `get_observatory_codes` method (not yet tested).  Further comments are welcome, but I'll chug along as I have time.

mkelley commented on astroquery/mpc/core.py:
```diff
@@ -303,5 +327,350 @@ def get_mpc_endpoint(self, target_type):
+        assert (isinstance(location, (str, int, EarthLocation))
+                or np.isiterable(location)), "location parameter must be a string, integer, iterable, or astropy EarthLocation"
+
+        assert isinstance(start, (type(None), Time, str)
```
Oh, that is good to know!

mkelley commented on astroquery/mpc/core.py:
```diff
@@ -303,5 +327,350 @@ def get_mpc_endpoint(self, target_type):
+                          ), "start must be a string, an astropy Time object, or None."
+
+        _step = u.Quantity(step)
+        assert _step.unit in [
```
I just tried that out and the query fails for integers above 240.  So we can't, for example, convert everything to seconds.  And it requires integers, so no converting seconds to other units.

mkelley commented:
One last thing to figure out: the MPES documentation says all dates are UT, but requesting heliocentric and geocentric vectors returns a column labeled "JD_TT", implying it is in the Terrestrial Time scale.  This code converts all input times to UTC, so if a user provides a Time in the TT scale, it will be converted to UTC before being passed to the MPES.  The returned dates will not be what the user requested.  I think this is primarily a documentation issue on the MPC's side.  Which is correct, the MPES docs or the returned table column heading JD_TT?  I'll ask the MPC for clarification.

Then time to write documentation.

mkelley commented:
OK, I;ve run into an issue.  My formatting of Angle columns is working well.  For example, we can format ra as degrees or hour angle, and vary the precision on the last decimal:
```
In [81]: astroquery.mpc.MPC.get_ephemeris('129P', start='2018-07-10', ra_format={'unit': 'deg', 'sep': ':', 'precision': 0})
Out[81]: 
<Table length=21>
          Date              RA       Dec     Delta     r    Elongation  Phase     V    Proper motion Direction
                           deg       deg       AU      AU      deg       deg     mag     arcsec / h     deg   
         object          float64   float64  float64 float64  float64   float64 float64    float64     float64 
----------------------- --------- --------- ------- ------- ---------- ------- ------- ------------- ---------
2018-07-10 00:00:00.000 322:05:36 -11:25:49   3.756   4.638      146.5     6.9    20.5         15.05     253.4
2018-07-11 00:00:00.000 321:59:37 -11:27:34   3.748   4.638      147.6     6.7    20.5         15.47     253.2
2018-07-12 00:00:00.000 321:53:30 -11:29:23    3.74   4.638      148.6     6.5    20.5         15.88     253.1
2018-07-13 00:00:00.000 321:47:13 -11:31:16   3.732   4.638      149.7     6.3    20.5         16.28     253.0
2018-07-14 00:00:00.000 321:40:48 -11:33:12   3.724   4.638      150.8     6.1    20.5         16.67     252.9
2018-07-15 00:00:00.000 321:34:13 -11:35:11   3.717   4.638      151.8     5.9    20.5         17.05     252.8
2018-07-16 00:00:00.000 321:27:30 -11:37:14    3.71   4.638      152.9     5.7    20.5         17.41     252.7
2018-07-17 00:00:00.000 321:20:39 -11:39:21   3.703   4.638      153.9     5.5    20.5         17.77     252.6
                    ...       ...       ...     ...     ...        ...     ...     ...           ...       ...
```
```
In [84]: tab = astroquery.mpc.MPC.get_ephemeris('129P', start='2018-07-10', ra_format={'unit': 'hourangle', 'sep': ':', 'precision': 1})

In [85]: tab
Out[85]: 
<Table length=21>
          Date              RA        Dec     Delta     r    Elongation  Phase     V    Proper motion Direction
                        hourangle     deg       AU      AU      deg       deg     mag     arcsec / h     deg   
         object          float64    float64  float64 float64  float64   float64 float64    float64     float64 
----------------------- ---------- --------- ------- ------- ---------- ------- ------- ------------- ---------
2018-07-10 00:00:00.000 21:28:22.4 -11:25:49   3.756   4.638      146.5     6.9    20.5         15.05     253.4
2018-07-11 00:00:00.000 21:27:58.5 -11:27:34   3.748   4.638      147.6     6.7    20.5         15.47     253.2
2018-07-12 00:00:00.000 21:27:34.0 -11:29:23    3.74   4.638      148.6     6.5    20.5         15.88     253.1
2018-07-13 00:00:00.000 21:27:08.9 -11:31:16   3.732   4.638      149.7     6.3    20.5         16.28     253.0
2018-07-14 00:00:00.000 21:26:43.2 -11:33:12   3.724   4.638      150.8     6.1    20.5         16.67     252.9
2018-07-15 00:00:00.000 21:26:16.9 -11:35:11   3.717   4.638      151.8     5.9    20.5         17.05     252.8
2018-07-16 00:00:00.000 21:25:50.0 -11:37:14    3.71   4.638      152.9     5.7    20.5         17.41     252.7
2018-07-17 00:00:00.000 21:25:22.6 -11:39:21   3.703   4.638      153.9     5.5    20.5         17.77     252.6
                    ...        ...       ...     ...     ...        ...     ...     ...           ...       ...
```

And this writes out to text files just fine:
```
In [88]: tab.write('/tmp/test.txt', format='ascii')
In [89]: cat /tmp/test.txt
Date RA Dec Delta r Elongation Phase V "Proper motion" Direction
"2018-07-10 00:00:00.000" 21:28:22.4 -11:25:49 3.756 4.638 146.5 6.9 20.5 15.05 253.4
"2018-07-11 00:00:00.000" 21:27:58.5 -11:27:34 3.748 4.638 147.6 6.7 20.5 15.47 253.2
"2018-07-12 00:00:00.000" 21:27:34.0 -11:29:23 3.74 4.638 148.6 6.5 20.5 15.88 253.1
"2018-07-13 00:00:00.000" 21:27:08.9 -11:31:16 3.732 4.638 149.7 6.3 20.5 16.28 253.0
"2018-07-14 00:00:00.000" 21:26:43.2 -11:33:12 3.724 4.638 150.8 6.1 20.5 16.67 252.9
"2018-07-15 00:00:00.000" 21:26:16.9 -11:35:11 3.717 4.638 151.8 5.9 20.5 17.05 252.8
"2018-07-16 00:00:00.000" 21:25:50.0 -11:37:14 3.71 4.638 152.9 5.7 20.5 17.41 252.7
```

The formatting is implemented with a lambda function for the `Column` format:
```
                # apply custom format
                tab['RA'].format = lambda ra: Angle(
                    ra, self._ra_format['unit']).to_string(**self._ra_format)
                tab['Dec'].format = lambda dec: Angle(
                    dec, self._dec_format['unit']).to_string(**self._dec_format)
```

However, writing to the ecsv format raises an exception because `yaml` cannot serialize the lambda function:
```
In [90]: tab.write('/tmp/test.csv', format='ascii.ecsv')
---------------------------------------------------------------------------
RepresenterError                          Traceback (most recent call last)
<ipython-input-90-276e79e47a52> in <module>()
----> 1 tab.write('/tmp/test.csv', format='ascii.ecsv')
...
~/local/miniconda3/envs/astroquery-dev/lib/python3.6/site-packages/yaml/representer.py in represent_undefined(self, data)
    227 
    228     def represent_undefined(self, data):
--> 229         raise RepresenterError("cannot represent an object: %s" % data)
    230 
    231 SafeRepresenter.add_representer(type(None),

RepresenterError: cannot represent an object: <function MPCClass._parse_result.<locals>.<lambda> at 0x7f5f83247d08>
```

As an alternative, I thought that we could use the `.dms` and `.hms` attributes on `Angle` but that runs into rounding issues: 
```
In [94]: '{0.dms.d:.0f}:{0.dms.m:.0f}:{0.dms.s:.1f}'.format(Angle(1, 'hr'))
Out[94]: '14:59:60.0'
```

I don't think it is a good idea to create tables that cannot be written to one of `astropy`'s premier table formats, which means I'll have to strip the `ra_format` and `dec_format` options.  I'll leave the columns as arrays of floats with the proper units, though.

mkelley commented on astroquery/mpc/tests/test_mpc_remote.py:
```diff
@@ -22,18 +26,22 @@ def test_query_object_valid_object_by_name(self, type, name):
         print(response.json())
-        print(mpc.core.MPC.query_object_async(get_query_payload=True, target_type=type, designation=designation))
+        print(mpc.core.MPC.query_object_async(get_query_payload=True,
+                                              target_type=type, designation=designation))
```
They must have been.  I'll remove them.

mkelley commented on astroquery/mpc/tests/test_mpc_remote.py:
```diff
@@ -1,8 +1,12 @@
 import pytest
 
+import astropy.units as u
+from astropy.time import Time
```
Removed.

mkelley commented on astroquery/mpc/tests/test_mpc_remote.py:
```diff
@@ -1,8 +1,12 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
+import numpy as np
```
Removed.

mkelley commented on astroquery/mpc/tests/test_mpc.py:
```diff
@@ -17,24 +80,241 @@ def patch_post(request):
+
+
+@pytest.mark.parametrize('perturbed,val', ((True, 'y'), (False, 'n')))
+def test_get_ephemeris_suppress_set(perturbed, val):
```
Fixed.

mkelley commented:
I think this is done, although I don't know how to preview the documentation.
* I only documented what I though was the most common usage.  Let me know if more should be added.
* The code in the MPC object query documentation did not work, so I fixed it.
* I decided to restore the `ra_format` and `dec_format` keywords.  I was documenting how to display coordinates as HH:MM:SS, etc., and decided making everyone else write this code over and over again was not a good solution.  When format keywords are used, the respective column is returned as a string (carrying the appropriate units), which will round-trip just fine with `Table.write`.
* I've added `get_observatory_location` to return the coordinates and name of a single observatory from the MPC list.

keflavich commented:
There appear to be some lingering problems with the docstrings:
https://travis-ci.org/astropy/astroquery/jobs/416862905#L2217

Also, JSONDecodeError is apparently py3-only:
https://travis-ci.org/astropy/astroquery/jobs/416862907#L3111

keflavich commented on astroquery/mpc/core.py:
```diff
@@ -274,34 +310,595 @@ def query_objects_async(self, target_type, get_query_payload=False, *args, **kwa
+        Returns
+        -------
+        tab : Table
+            Table of codes, coordinates, and names.
```
This is actually incorrect: the `async` method is returning a `requests` response object, not a Table.  The docstring failure is happening because the `async_to_sync_docstr` processor is failing here... https://github.com/astropy/astroquery/blob/master/astroquery/utils/process_asyncs.py#L56

keflavich replied:
Actually, I'm not certain the failure is related to this, but I suspect so still...

mkelley replied:
textwrap.dedent(doc) did not work as expected on this docstring because I've always started my docstrings just after the triple quote, i.e., not indented!  Because of this the Returns section was not stripped and the new one appended, resulting in two Returns sections and the error.

I'll fix the async section and my docstrings then I think this will pass.

migueldvb commented on docs/mpc/mpc.rst:
```diff
@@ -29,68 +42,293 @@ An example of an exact match:
+Makemake's ephemeris for the Discovery Channel Telescope (IAU
+observatory code G37):
+
+    >>> eph = MPC.get_ephemeris('Makemake', location='G37')
```
This needs the `.. code-block:: python` directive

mkelley replied:
Got it.

mkelley commented:
I'm working on understanding all the warnings from the docs.

mkelley commented:
I'm not sure what to do here. I have a table with asteroid designations in it:
```
            +------------+-----------------------------------+
            | Target     | Description                       |
            +============+===================================+
            | \(3202\)   | Numbered minor planet (3202)      |
            +------------+-----------------------------------+
            | 14829      | Numbered minor planet (14829)     |
            +------------+-----------------------------------+
```
reST will interpret the `(3202)` as the 3202nd item in a list.  To prevent that, I added the escape characters.  I think Travis doesn't like it:
```
  File "/home/travis/build/astropy/astroquery/astroquery/mpc/core.py", line 534
            """
SyntaxError: invalid escape sequence \(
```
and
```
coverage.misc.NotPython: Couldn't parse '/home/travis/build/astropy/astroquery/astroquery/mpc/core.py' as Python source: 'invalid escape sequence \(' at line 534
```

keflavich commented:
I think this is a python3.6 thing.  I've had all sorts of issues with disallowed escape sequences raising errors, and I still don't really understand them.  But, try double escaping, `\\`?

mkelley commented:
Making the docstring a raw string avoids the coverage error.  I'm not satisfied with having to use the escape character though.  I added a note in the documentation indicating that the \ should be ignored.

mkelley commented:
How do I address the python 2.7 build stalling in Travis?

keflavich commented:
No idea.  I restarted it, but that's unusual.

migueldvb commented on astroquery/mpc/tests/test_mpc.py:
```diff
@@ -16,25 +84,276 @@ def patch_post(request):
+
+
+def test_get_ephemeris_Moon_phase():
+    result = mpc.core.MPC.get_ephemeris('2P', location='G37')
```
I think that `proper_motion='coordinate'` is needed in the call to get_ephemeris to match the existing data file

mkelley replied:
The default is total, so I have the wrong file names!

migueldvb commented on astroquery/mpc/tests/test_mpc.py:
```diff
@@ -16,25 +84,276 @@ def patch_post(request):
+    assert result['Moon phase'][0] >= 0
+
+
+def test_get_ephemeris_Uncertainty():
```
`patch_post` is needed here and a a few other functions

mkelley commented on astroquery/mpc/tests/test_mpc.py:
```diff
@@ -16,25 +84,276 @@ def patch_post(request):
+    return os.path.join(data_dir, filename)
+
+
 def post_mockreturn(self, httpverb, url, params, auth):
```
Only the ephemeris service is POST.  The object query and observatory code requests are GET.

mkelley replied:
OK, so I'll define get_mockreturn for the gets :)

mkelley commented on astroquery/mpc/tests/test_mpc.py:
```diff
@@ -16,25 +84,276 @@ def patch_post(request):
+        prefix = params['TextArea'].replace(' ', '')
+        suffix = '-'.join((params['c'], params['raty'], params['s']))
+        filename = data_path('{}_ephemeris_{}.html'.format(prefix, suffix))
+        content = open(filename, 'r').read()
```
Oh, I hadn't realized that.  I guess the same for the other files, too.

mkelley commented:
Travis stopped with a network error.  Can I restart this, or does it take a maintainer to do that?

pllim commented:
@mkelley , I restarted it for you. Theoretically you should be able to restart it. But if you don't see the option or if it does not work as advertised, feel free to ping a maintainer to do it for you. Thanks!

bsipocz commented:
@pllim - I believe it's only the maintainers (either astroquery, and a subset of astropy) who can restart builds.

@mkelley - feel free to ping us on the #ci-issues channel on slack, that may trigger a quicker response.

mkelley commented:
All checks pass!  @keflavich @bsipocz @migueldvb ready to merge?

mkelley commented:
Thanks everyone!

bsipocz commented:
thank you @mkelley!
